cf_xarray/accessor.py

```diff
@@ -2254,23 +2268,51 @@ def decode_vertical_coords(self, prefix="z"):
                     terms["depth_c"] * terms["s"]
                     + (terms["depth"] - terms["depth_c"]) * terms["C"]
                 )
+                # # expand dims so ordering is preserved
+                # terms["eta"] = terms["eta"].expand_dims(
+                #     dim={terms["s"].name: terms["s"]}, axis=1
+                # )
                 # z(n,k,j,i) = S(k,j,i) + eta(n,j,i) * (1 + S(k,j,i) / depth(j,i))
-                ds.coords[zname] = S + terms["eta"] * (1 + S / terms["depth"])
+                ztemp = terms["eta"] * (1 + S / terms["depth"]) + S
 
             elif stdname == "ocean_s_coordinate_g2":
                 # make sure all necessary terms are present in terms
                 # (depth_c * s(k) + depth(j,i) * C(k)) / (depth_c + depth(j,i))
                 S = (terms["depth_c"] * terms["s"] + terms["depth"] * terms["C"]) / (
                     terms["depth_c"] + terms["depth"]
                 )
+                # # expand dims so ordering is preserved
```

Comment:
I don't see how this could work.

```
# S(k,j,i) + eta(n,j,i) * (1 + S(k,j,i) / depth(j,i))
```
 We want this to work even for subsets along a S-level (k=0) or time (n=0). In which case, inserting a dimension at axis=1 for eta would totally break (expand_dims with a scalar will use that scalar as size, or just error for negative value of s)

I experimented with explicit broadcasting but can't get that to work either.

```
# S(k,j,i) + eta(n,j,i) * (1 + S(k,j,i) / depth(j,i))
S = xr.DataArray(np.ones((4, 3, 2)), dims=("z", "y", "x"))
eta = xr.DataArray(np.ones((5, 3, 2)), dims=("t", "y", "x"))
depth = xr.DataArray(np.ones((3, 2)), dims=("y", "x"))
xr.broadcast(S, eta)
```

Reply:
I think if we could count on dimensions being properly identified this would be straight-forward: the calculation could add singleton dimensions for any missing dimensions, do the calculation, the order would be proper, and then we could squeeze at the end to return. But, I don't think we can count on it working generally. So, I think we should remove all this attempt at reordering and just stick with the new calculation. I'll take it out and let me know if you think otherwise.

Reply:
:+1: I can't see how we make it work in general without specific knowledge of the actual model variable names. Thanks for checking!